[WIP] Tpetra: Enhancement to Tpetra::Details::Behavior

tjfulle opened the pull request:
@mhoemmen, this PR outlines my idea on the enhancement to the `Tpetra::Details::Behavior` class to be able to conditionally specify behavior on a per class/object basis.  I implemented the enhanced `Behavior` methods and adopted them in `Tpetra::Distributor`.  On the test problem I was working on for #1752, it worked beautifully to run my example executable as

```
TPETRA_VERBOSE=Distributor ./executable-name
```

when I wanted verbose output, but only for `Distributor`.

There are probably optimizations to be made, but I think this would be a really nice addition.

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -22,6 +24,89 @@ namespace { // (anonymous)
 
+  template<class Func>
+  void
+  split(const std::string& s, Func f, const char sep=':')
```
You could hide the implementation of `split`, by using `std::function<void(const std::string&)>` as the type of `f`, instead of templating on `Func`.  This would also better express intent, without restricting the interface in any way (the code should be equivalent).

mhoemmen replied:
I'm getting super into implementation hiding as a result of working with @prwolfe :-) .

mhoemmen replied:
Another thing: If `:` is the separator, then what do we do about classes in the `Tpetra::Details` namespace?  I'm OK with just ignoring them, but just want to know what you think.

tjfulle replied:
I chose `:` mainly for its use as the posix pathsep character.  The name passed to `verbose` or `debug` doesn't necessarily have to be fully namespace qualified.

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -22,6 +24,89 @@ namespace { // (anonymous)
+  /// If the environment variable is "false", return 0
+  /// Otherwise, return -1
+  int
+  environmentVariableState(std::string environmentVariableValue)
```
Recommend passing `std::string` as `const std::string&`.

tjfulle replied:
yep, it should be `const std::string&`

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -22,6 +24,89 @@ namespace { // (anonymous)
+    // Variable is not empty.
+    const string varStr(varVal);
+    std::vector<string> names;
+    split(varStr, [&](string x){names.push_back(x);});
```
It would be nicer to pass `string` as `const string&` here.

tjfulle replied:
ditto

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -22,6 +24,89 @@ namespace { // (anonymous)
+    const string varStr(varVal);
+    std::vector<string> names;
+    split(varStr, [&](string x){names.push_back(x);});
+    for (auto name: names) {
```
You could do this, to avoid `std::string` construction:
```
for (auto&& name : names) { /* ... */ }
```

tjfulle replied:
ditto

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -22,6 +24,89 @@ namespace { // (anonymous)
+
+  /// \brief Determine if environment variable is a boolean or not
+  ///
+  /// If the environment variable is "true", return 1
```
An `enum` might be nicer here, but this is fine for now.

tjfulle replied:
`enum` where?

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -22,6 +24,89 @@ namespace { // (anonymous)
+    for (auto name: names) {
+      int state = environmentVariableState(name);
+      if (state == 1) {
+        // Global value takes precedence
```
I would say, the global value is the default value for all classes.  However, users may still turn on or off the value for particular classes.  This would imply returning the global default value separately, not as part of the `std::map`.

tjfulle replied:
The behavior is almost like you suggest already.  The only case where there is ambiguity is if, say `TPETRA_VERBOSE` is specified as:

```TPETRA_VERBOSE=Distributor:CrsMatrix:ON```

The `ON` value would trigger `VERBOSE` to be `ON` for all classes.

mhoemmen replied:
Ah, OK then.  That's fine.  Silly of the user, but fine ;-) .

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -22,6 +24,89 @@ namespace { // (anonymous)
+        break;
+      }
+      else if (state == 0) {
+        // Global value takes precedence
```
See above comment (global value is the default for all classes; class-specific values should take precedence over that).

tjfulle replied:
see my response above.  I could refactor it so that the map *only* stores values for class-specific cases.  Then, the `verbose` (or `debug`) functions would return the default if a class-specific value is not found in the map.  I'm not sure that made sense, but I like it better, I'll throw that in.

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -84,6 +163,47 @@ namespace { // (anonymous)
+          initialized = true;
+        });
+    }
+    return;
```
You don't need this `return` statement.

tjfulle replied:
okay

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -84,6 +163,47 @@ namespace { // (anonymous)
   }
 
+  void
+  idempotentlyGetEnvironmentVariableAsMap (std::once_flag& once_flag,
```
Since `valsMap won't get filled in if this function has been called before, it might be better to call this `getEnvironmentVariableAsMapOnce`.  (It's not technically "idempotent" if calling it the second time gives a different result than calling it the first time.)  This is just a quibble over naming, not a correctness issue (but see comments below).

tjfulle replied:
okay

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -124,6 +244,50 @@ bool Behavior::verbose ()
+{
+  static std::once_flag debugFlag_;
+  static bool debugInitialized_ = false;
+  static std::map<std::string, bool> namedDebugValues_;
```
Putting initialization of the `static std::map` here, instead of inside the function to call via `std::call_once`, breaks reëntrancy of `Behavior::debug`.  Imagine if two different threads enter `Behavior::debug` concurrently.  They may both invoke the constructor of `std::map` concurrently, which could cause troubles (note that the default constructor doesn't promise not to allocate memory; for example, `std::vector` may reserve space for a small number of entries by default, if not given an initial number of entries).

tjfulle replied:
I see.  How would `namedDebugValues_` be declared then?  A pointer?

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -124,6 +244,50 @@ bool Behavior::verbose ()
+{
+  static std::once_flag verboseFlag_;
+  static bool verboseInitialized_ = false;
+  static std::map<std::string, bool> namedVerboseValues_;
```
See above comment in `Behavior::debug`.

tjfulle commented:
Okay @mhoemmen, I incorporated your suggestions.  The `std::map` is still not quite right.

tjfulle commented:
@mhoemmen, okay, now the `std::map` lives in its own namespace as you suggested.

mhoemmen commented:
@tjfulle I'll take a look.

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -22,6 +37,81 @@ namespace { // (anonymous)
 
+  void
+  split(const std::string& s,
+        std::function<void(const std::string&)> f,
```
I forgot to mention -- you have to include `<functional>` in order to use `std::function`.

tjfulle replied:
hmm, built and passed tests on my machine.  but, i'll add it

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -84,6 +168,51 @@ namespace { // (anonymous)
+          // read or write which precedes it in program order with any
+          // write which follows it in program order."
+          //
+          // The point is to prevent the assignment to 'value' from
```
Be sure to fix the comment :)

mhoemmen commented:
@tjfulle Alas, my #1860 fix caused some conflicts in `Tpetra::Details::Behavior`....

tjfulle commented:
@mhoemmen @kddevin, is this a capability that others would find useful?  I found it extremely useful when debugging a single Tpetra class/function and not wanting verbose Tpetra output globally.  If it indeed would be useful to others, I could resolve conflicts and have this PR ready to merge in very little time (minutes?).  If not, I'd be inclined to remove the PR rather than have it sit in a backlog queue where merge conflicts could potentially be much larger.

mhoemmen commented:
@tjfulle We (@kddevin , @crtrott , and I) actually discussed this today.  I said that I would like this feature and would be happy to accept, pending resolution of merge conflicts.  If you feel like fixing the merge conflicts, I'll be happy to accept immediately.  Thanks!

tjfulle commented:
Thanks @mhoemmen !  I guess I wrongly assumed the classification as "backlog" meant something that may or may not ever get looked at.  I'll resolve the conflicts and get the PR updated

mhoemmen commented:
@tjfulle "backlog" just means "haven't started yet" :-)

tjfulle commented:
@mhoemmen, I fixed merged conflicts and a correctness issue with the named variants of the `Behavior` methods.  I also added unit tests for all of the core `Behavior` functionality.  When you have a moment, could you look it over again.  If you approve, I'll remove the [WIP] designator and run the full checkin test.

Semi related, since the `Behavior` class supersedes the `Tpetra::Details::Environment` class, I think we should yank `Tpetra::Details::Environment`.  I could do that as part of this PR.

mhoemmen commented:
Hi @tjfulle !  Thanks for working on this!  I'll take a look.

> Semi related, since the `Behavior` class supersedes the `Tpetra::Details::Environment` class, I think we should yank `Tpetra::Details::Environment`. I could do that as part of this PR.

I agree we should yank it.  Sorry we ended up not using it!  It doesn't have to be part of this PR; it can happen separately.  Up to you :)

tjfulle commented:
> I agree we should yank it. Sorry we ended up not using it! It doesn't have to be part of this PR; it can happen separately. Up to you :)

@mhoemmen, I went ahead and yanked it.  The PR now reflects the updated `Behavior` and removed `Environment`

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -84,6 +175,50 @@ namespace { // (anonymous)
+          initialized = true;
+        });
+    }
+    auto thisEnvironmentVariableMap = namedVariableMap_[environmentVariableName];
```
This _should_ be a read-only access, because `setEnvironmentVariableMap` should have already added the `environmentVariableName` key to `namedVariableMap_`.  I'm not sure whether `operator[]` and `find` have the same thread-safety guarantees for read-only access, but I think it's fine for now.

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.cpp:
```diff
@@ -84,6 +175,50 @@ namespace { // (anonymous)
+    }
+    auto thisEnvironmentVariableMap = namedVariableMap_[environmentVariableName];
+    if (thisEnvironmentVariableMap.find(name) != thisEnvironmentVariableMap.end())
+      return thisEnvironmentVariableMap[name];
```
`find` looks up the key; `operator[]` does it again.  Try this:
```
auto tevmIter = thisEnvironmentVariableMap.find (name);
if (temvIter != thisEnvironmentVariableMap.end ()) {
  return tevmIter->second;
}
```

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.hpp:
```diff
@@ -58,13 +78,15 @@ class Behavior {
   /// also produce more detailed error messages, and more copious
   /// debug output.
   static bool debug ();
+  static bool debug (const char name[]);
```
Add the following comment:
```
/// \brief Whether the given Tpetra class is in debug mode.
///
/// \param name [in] Name of the Tpetra class: e.g., "CrsGraph"
```

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_Behavior.hpp:
```diff
@@ -58,13 +78,15 @@ class Behavior {
   /// std::cerr on every MPI process.  This is a LOT of output!  You
   /// really don't want to do this when running on many MPI processes.
   static bool verbose ();
+  static bool verbose (const char name[]);
```
Add the following comment:
```
/// \brief Whether the given Tpetra class is in verbose mode.
///
/// \param name [in] Name of the Tpetra class: e.g., "CrsGraph"
```

mhoemmen commented on packages/tpetra/core/src/Tpetra_Distributor.cpp:
```diff
@@ -159,6 +160,8 @@ namespace Tpetra {
         (out_.is_null (), std::logic_error, "Tpetra::Distributor::init: debug_ "
```
Line 158: `if (debug_)` needs to read `if (verbose)`.  Ditto for the error message.  The issue is that the error messages need to print based on `verbose`.

mhoemmen replied:
^^^ this is the only actual bug I found ^^^

tjfulle commented:
@mhoemmen, thanks for the feedback!  I addressed each comment.  Commit message has been updated to reflect that all tests now pass with the checkin script.

tjfulle commented on packages/tpetra/core/src/Tpetra_Distributor.cpp:
```diff
@@ -154,11 +155,13 @@ namespace Tpetra {
 
-    if (debug_) {
+    if (verbose || debug_) {
       TEUCHOS_TEST_FOR_EXCEPTION
```
@mhoemmen, I made the check for `verbose` or `debug_`.  I imagine it is conceivable that if `debug_` one would want to write to `out_` (though the preferred would be to write to `out_` only if `verbose`)

mhoemmen replied:
@tjfulle Does `out_` get initialized in either case?  If so, then I'm OK with this.

tjfulle replied:
@mhoemmen, according to [Line 148 of Tpetra_Distributor.cpp](https://github.com/tjfulle/Trilinos/blob/1e78c86eb2313a8535594f4c47a6d5ed785535cb/packages/tpetra/core/src/Tpetra_Distributor.cpp#L148), `out_` is set independent of `debug_` or `verbose` - to either the incoming value of `out` or `Teuchos::getFancyOStream (Teuchos::rcpFromRef (std::err))` if `out` is null.

mhoemmen commented:
@tjfulle As long as the `if (verbose || debug_)` works correctly in either case (with respect to initialization of `out_`), then I'm OK with this; feel free to test and push.  Thanks!

tjfulle commented:
@mhoemmen, according to [Line 148 of Tpetra_Distributor.cpp](https://github.com/tjfulle/Trilinos/blob/1e78c86eb2313a8535594f4c47a6d5ed785535cb/packages/tpetra/core/src/Tpetra_Distributor.cpp#L148), `out_` is set independent of `debug_` or `verbose` - to either the incoming value of `out` or `Teuchos::getFancyOStream (Teuchos::rcpFromRef (std::err))` if `out` is null.  The last commit was tested and pushed from the checkin script.  I'll squash and merge in the morning.

mhoemmen commented:
@tjfulle Cool, thanks for looking :-)  Thanks for working on this!
